rust/cardano-blockchain-types/src/cip36/key_registration.rs

```diff
@@ -0,0 +1,360 @@
+//! CIP-36 Key Registration 61284.
+//!
+//! Catalyst registration data
+//!
+//! <https://cips.cardano.org/cip/CIP-36>
+//! <https://github.com/cardano-foundation/CIPs/blob/master/CIP-0036/schema.cddl>
+
+use std::collections::HashSet;
+
+use ed25519_dalek::VerifyingKey;
+use minicbor::{decode, Decode, Decoder};
+use pallas::ledger::addresses::{Address, ShelleyAddress};
+use strum::FromRepr;
+
+use super::voting_pk::VotingPubKey;
+use crate::utils::decode_helper::{decode_array_len, decode_bytes, decode_helper, decode_map_len};
+
+/// CIP-36 key registration - 61284
+///
+///
+/// ```cddl
+/// key_registration = {
+///     1 : [+delegation] / legacy_key_registration,
+///     2 : $stake_credential,
+///     3 : $payment_address,
+///     4 : $nonce,
+///     ? 5 : $voting_purpose .default 0
+//   }
+/// ```
+#[allow(clippy::module_name_repetitions)]
+#[derive(Clone, Default, Debug)]
+pub struct Cip36KeyRegistration {
+    /// Is this CIP36 or CIP15 format.
+    #[allow(clippy::struct_field_names)]
+    pub is_cip36: Option<bool>,
+    /// Voting public keys (called Delegations in the CIP-36 Spec).
+    /// Field 1 in the CIP-36 61284 Spec.
+    pub voting_pks: Vec<VotingPubKey>,
```

Comment:
As far as I understand, the delegations array (`voting_pks` field) must not be empty. Also there are checks for duplicate keys, but we probably should also check that the mandatory fields are present during the decoding and not just use default values if they aren't.

Reply:
Agree that we should check for mandatory fields - added.

My thought is to separate between decoding from validation. Just feel like decoding should focus solely on extracting and interpreting the data. We can discuss it when @stevenj come back